Exercise_3.py

```diff
@@ -1,26 +1,240 @@
-# Node class  
-class Node:  
-  
-    # Function to initialise the node object  
-    def __init__(self, data):  
-        
-class LinkedList: 
-  
-    def __init__(self): 
-        
-  
-    def push(self, new_data): 
-        
-  
-    # Function to get the middle of  
-    # the linked list 
-    def printMiddle(self): 
-
-# Driver code 
-list1 = LinkedList() 
-list1.push(5) 
-list1.push(4) 
-list1.push(2) 
-list1.push(3) 
-list1.push(1) 
-list1.printMiddle() 
+'''2 Approaches'''
+'''
+Algorithm (Fylod's Algorithm aka Tortoise and Hare Algorithm)
+    printMiddle(arr, low, high):
+        1. If there are no elements in the linked list return None
+        2. Initialize first and second pointer to head
+        3. If second and secnd.next are not None continue as go to step 6
+        4. For second take 2 steps by setting it to second.next.next
+        5. For first take 1 steps by setting it to first.next. Thus the second pointer
+           will moveat twice the speed of first one. So if it reaches the end of the list 
+           first will be at the middle of the list
+        6. Print first.data
+
+LinkedList
+    Space Complexity: space complexity of the node * no of elements in linked list = O(2) * O(n) = O(2n) = O(n)
```

Comment:
SC will be O(1) because without giving the linkedlist in input you can't find the middle